### PR TITLE
Add Retry Foundation - RequestType, RetryPolicy, and RetryTimeoutManager

### PR DESCRIPTION
**Pull request by nishkarsh-db**
## Description

Introduces foundational components for unified retry handling mechanism:

- **RequestType enum**: Categorizes database operations (OpenSession, ExecuteStatement, etc.) with associated retry policies
- **RetryPolicy enum**: Defines IDEMPOTENT and NON_IDEMPOTENT policies
- **RetryTimeoutManager**: Manages separate timeout budgets for different error types:
  - Service unavailable (503) - configurable via `TemporarilyUnavailableRetryTimeout`
  - Rate limiting (429) - configurable via `RateLimitRetryTimeout`
  - API retriable codes - configurable via `ApiRetryTimeout`
  - Other HTTP errors - 10s default
  - Exceptions - 10s default
- **RetryUtils enhancement**: Added `extractRetryException()` method and timeout constants

This PR is the first in a stack of 4 PRs that will migrate retry handling from Apache HTTP client's native handlers to application-layer control.

## Testing

All tests pass (8 tests): `[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0`

`RetryTimeoutManagerTest` covers:
- Timeout exhaustion for each error type
- Mixed retry scenarios
- Edge cases (immediate exhaustion)

## Notes for Reviewer

- No breaking changes - only adds new components
- Subsequent PRs will integrate these components into HTTP (PR3) and Thrift (PR4) clients

### CODE REVIEW

File: src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryTimeoutManager.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * Evaluates retry decision based on HTTP status code and updates timeout accordingly. Uses the
+   * Retry-After header value when provided by the strategy.
+   *
+   * @param statusCode the HTTP status code from the response
+   * @param retryDelayMillis the retry delay in milliseconds to subtract from timeout
+   * @param isApiRetriableCode true if this is a custom API retriable code, false otherwise
+   * @return true if the request should be retried, false otherwise
+   */
+  public boolean evaluateRetryTimeoutForResponse(
```
Comment (vikrantpuppala) [inline src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryTimeoutManager.java]: can we add debug logs so that it is abundantly clear why a retry was stopped

File: src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryUtils.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  /**
+   * Extracts DatabricksRetryHandlerException from the exception cause chain.
+   *
+   * @param e the exception to search through
+   * @return the DatabricksRetryHandlerException if found, null otherwise
+   */
+  public static DatabricksRetryHandlerException extractRetryException(Throwable e) {
+    Throwable cause = e.getCause();
+    while (cause != null) {
```
Comment (vikrantpuppala) [inline src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryUtils.java]: do we want to skip top level e?

File: src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryTimeoutManager.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        break;
+      default:
+        otherErrorCodesTimeoutMillis -= retryDelayMillis;
+        break;
+    }
+
+    // Check if any timeout has been exceeded
+    return tempUnavailableTimeoutMillis > 0
+        && rateLimitTimeoutMillis > 0
+        && otherErrorCodesTimeoutMillis > 0;
```
Comment (vikrantpuppala) [inline src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryTimeoutManager.java]: can we just move these checks to the individual case in the switch block

File: src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryUtils.java (hunk 2 above)
Comment (nishkarsh-db) [reply src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryUtils.java]: Tes as it would always be of the type TTransportException, so we do want to skip it, I'll add a comment for that.

File: src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryTimeoutManager.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * Evaluates retry decision based on an exception and updates timeout accordingly.
+   *
+   * @param exception the exception that occurred during request execution
```
[the remark this answers is not in the document]
Comment (nishkarsh-db) [reply src/main/java/com/databricks/jdbc/dbclient/impl/http/RetryTimeoutManager.java]: Yes, passing it is a part of next PR.
